src/content/docs/cheerp/guides/clientlib.md

```diff
@@ -0,0 +1,327 @@
+
+## What is `_Any`, `_Union`, and `_Function`?
+
+These are 3 new types to help improve the type safety of programs that interact
+with javascript code. These types do not exist in javascript, they only exist
+to help define accurate function signatures.
+
+`_Any` is a lot like `Object`, except that `_Any` is also compatible with
+primitive types, such as `double` and `int`. `Object` instances are passed by
+pointer, and need to be constructed explicitly. `_Any` is passed by const
+reference and can be constructed on the fly for any type. `_Any` and `Object`
+are roughly equivalent to `any` and `object` in typescript, and should be used
+similarly.
+
+`_Union` is used to describe values whose type is unknown, but limited to a set
+of known types. `_Union` types are passed by const reference and can be
+constructed on the fly for any type in the union. `_Union` types can also be
+returned from functions, and cast to a real type using `->cast()` or
+`->cast<T>()`. `_Union<T...>` is roughly equivalent to `T | ...` in typescript,
+and should be used similarly.
+
+`_Function` is a type safe first-class function type that replaces
+`EventListener*` when passing functions to higher order functions. `_Function`
+types are passed by const reference and can be constructed on the fly from raw
+function pointers and C++11 lambdas. `cheerp::Callback()` can still be used to
+get an `EventListener*`, but prefer constructing a `_Function` instead.
+
+`_Union` and `_Function` support many conversions to help keep your code clean.
+For example, `_Union<double>` can be converted to `_Union<double, String*>`,
+and `_Function<void(Object*)>` can be converted to `_Function<void(String*)>`.
+
+## What is `->cast<T>()` and `->cast()`?
+
+For all all intents and purposes, these functions do *nothing*. They only
+change the type of a value. These differ from other pointer casts because they
+mimic the rules of casting objects in typescript, rather than in C++.
+
+`static_cast` has many restrictions that make in unsuitable for use with client
+types. For example, it does not allow you to cast through virtual base classes,
+which are used to define some client types. `static_cast` also does not allow
+you to perform many casts involving `_Union` and `_Function`. Because, in the
+eyes of `static_cast`, `_Union<String*>` and `_Union<Object*>` are two
+completely distinct and unrelated types. Using `->cast<T>()` and `->cast()`,
+you can perform these casts.
+
+`->cast<T>()` does exactly what you'd expect, it returns a value of type `T`.
+`->cast()` returns an intermediate object that implements a conversion operator
+`operator T();`. In many cases, the compiler can infer the type `T` of the
+conversion operator, and it doesn't need to be specified explicitly.
+
+```cpp
+    // auto cannot be inferred
+    auto* string = object->cast<client::String*>();
+
+    // type can be inferred.
+    client::String* string = object->cast();
+```
+
+## Unable to solve a migration issue?
```

Comment:
I don't like having this heading be a question when the other headings are not. Perhaps `Ask for help`